src/Marten/Linq/CollectionAnyContainmentWhereFragment.cs

```diff
@@ -112,5 +130,78 @@ private static void gatherSearch(BinaryExpression x, Dictionary<string, object>
                 throw new NotSupportedException();
             }
         }
+
+        private string buildSubQuery(SubQueryExpression subQuery, NpgsqlCommand command)
+        {
+            var contains = subQuery.QueryModel.ResultOperators.OfType<ContainsResultOperator>().FirstOrDefault();
+            if (contains == null)
+            {
+                throw new NotSupportedException("Only the Contains() operator is supported on subqueries within Collection.Any() searches");
+            }
+
+            // build rhs of ?|
+            var from = subQuery.QueryModel.MainFromClause.FromExpression as ConstantExpression;
+            if (from == null || !supportedTypes.Any(supp => isListOrArrayOf(from.Type, supp)))
+            {
+                throwNotSupportedContains();
+            }
+            var fromParam = command.AddParameter(from.Value);
+            fromParam.NpgsqlDbType = NpgsqlDbType.Array | NpgsqlDbType.Text;
+
+            // check/build lhs of ?|
+            var item = contains.Item as QuerySourceReferenceExpression;
+            if (item == null)
+            {
+                throwNotSupportedContains();
+            }
+            if (!supportedTypes.Any(supp => supp == item.ReferencedQuerySource.ItemType))
+            {
+                throwNotSupportedContains();
+            }
+            var itemSource = item.ReferencedQuerySource as MainFromClause;
+            if (itemSource == null)
+            {
+                throwNotSupportedContains();
+            }
+            var member = itemSource.FromExpression as MemberExpression;
+            if (member == null)
+            {
+                throwNotSupportedContains();
+            }
+
+            return $"data->'{pathTo(member)}' ?| :{fromParam.ParameterName}";
+        }
+
+        private void throwNotSupportedContains()
+        {
+            throw new NotSupportedException($"The Contains() operator on subqueries within Collection.Any() searches only supports constant array/lists of {string.Join(" or ", supportedTypes.Select(t => t.Name))} expressions");
+        }
+
+        private string pathTo(MemberExpression memberExpression)
+        {
+            var path = new List<string> { memberExpression.Member.Name };
+            while (memberExpression.Expression.NodeType == ExpressionType.MemberAccess)
+            {
+                var propInfo = memberExpression.Expression.GetType().GetProperty("Member");
+                var propValue = propInfo.GetValue(memberExpression.Expression, null) as PropertyInfo;
+                path.Add(propValue.Name);
+                memberExpression = memberExpression.Expression as MemberExpression;
+            }
+            return path.AsEnumerable().Reverse().Join("'->'");
+        }
+
+        private bool isListOrArrayOf(Type value, Type valid)
```

Comment:
I took a quick search and could not find anything suitable. If anyone knows a better way to check that, I'll be glad to change it.